run/block.ts

```diff
@@ -1,174 +1,154 @@
 import type {
   Block,
-  Exhausted,
-  Exited,
+  BlockResult,
   Frame,
   Instruction,
-  IterationEvent,
   Operation,
   Result,
 } from "../types.ts";
-import type { Computation } from "../deps.ts";
 
 import { createObservable } from "./observer.ts";
 import { evaluate, reset, shift } from "../deps.ts";
 import { lazy } from "../lazy.ts";
 import { create } from "./create.ts";
 
+type InstructionResult = {
+  type: "settled";
+  result: Result<unknown>;
+} | {
+  type: "interrupted";
+};
+
 export function createBlock<T>(
   frame: Frame,
   operation: () => Operation<T>,
 ): Block<T> {
-  let exhaustion: Exhausted<T> | undefined = void 0;
-  let observable = createObservable<IterationEvent<T>>();
+  let result: BlockResult<T> | void = void 0;
+  let results = createObservable<BlockResult<T>>();
+  let interruption = createObservable<void>();
+  let queue = createObservable<ReturnType<typeof $next>>();
   let controller = new AbortController();
-  let { notify } = observable;
+  let { signal } = controller;
+
+  signal.addEventListener("abort", () => interruption.notify());
 
   let enter = evaluate<() => void>(function* () {
     yield* shift<void>(function* (k) {
       return k;
     });
 
-    let iterator = lazy(() => operation()[Symbol.iterator]());
+    let thunks = queue.observe();
 
-    let exit = yield* shift<Exited<T>>(function* (k) {
-      let { signal } = controller;
-      signal.onabort = () => {
-        k({
-          type: "exited",
-          reason: "terminated",
-          result: { type: "rejected", error: new Error("halted") },
-        });
-      };
-
-      k({
-        type: "exited",
-        reason: "completed",
-        result: yield* reduce({
-          frame,
-          iterator,
-          notify,
-          signal,
-          start: $next(void 0),
-        }),
-      });
+    yield* reset(function* () {
+      result = yield* results.first();
```

Comment:
> With this code, there's nothing stopping us from erroneously emitting multiple times and no safe-guards to prevent it.

I see your point in that it does feel like a strange overload, however after the first event is received, the observer will be dropped, the continuation will be resumed, and because it is not in a while loop, subsequent events will not trigger anything.... If an event is dispatched, but there are no observers to consume it, did it happen?

Still, I share your feeling of counterintution regarding an event stream that we want for a single value. I have been thinking that maybe what we want is to make `Observable<T>` an `EventStream<T,TClose>`. And it would have a `close(value: TClose)` method to set the return state. Once the event stream is closed, any subsequent subscribers would immediately receive the end state. That would be a pretty elegant way to model the result of a block or frame, and it would prevent a lot of code like you see with block:

```ts
if (result) {
  return result;
} else {
  return yield* results.first();
}
```

Instead it would just be:

```ts
return yield* forEach(results);
```

I'm imagining `forEach` would be a way to consume an event stream that invoked a computation for each event in an EventStream, and return the `TClose` value as its own return value.

```ts
function forEach<T, TClose>(stream: EventStream<T,TClose>, each?: (t: T) => Computation<void>): Computation<TClose> {
  let observer = stream.observe();
  while (true) {
    let next = yield* observer;
    if (next.done) { 
      return next.value 
    } else if (each) {
      yield* each(next.value);
    }
  }
}
```

Reply:
I like the name change, Observable is an overloaded term in JS.

My comment about emitting multiple times with nothing listening to it was mainly out of caution.  We don't want to be in a situation where we are trying to debug our event streams, scratching our heads for why `notify()` wasn't doing what we expected, only to learn that all of our consumers were already consumed.  Basically, `notify()` implicitly turns into a no-op, where I'd almost rather see it through an error or warning, but maybe that's not possible.

I think your suggested changes make sense, thanks!